File: Advanced/PageAndRDL/CustomPdfExport/C#/PdfRendering/Pdf/PdfSharpRenderingExtension.cs
using System;
using System.Collections.Specialized;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.Extensibility.Rendering.Components;
using GrapeCity.ActiveReports.Extensibility.Rendering.IO;
using GrapeCity.ActiveReports.Rendering.Export;
using GrapeCity.ActiveReports.Samples.Export.Rendering.PdfSharp;

namespace GrapeCity.ActiveReports.Samples.Export.Rendering.Pdf
{
	/// <summary>
	/// <see cref="IRenderingExtension" /> implementation used to create a PDF.
	/// </summary>
	public sealed class PdfSharpRenderingExtension : IRenderingExtension, IConfigurable
	{
		#region IRenderingExtension implementation

		/// <summary>
		/// Renders report.
		/// </summary>
		void IRenderingExtension.Render(IReport report, StreamProvider streamProvider)
		{
			((IRenderingExtension)this).Render(report, streamProvider, null);
		}

		/// <summary>
		/// Renders report.
		/// </summary>
		void IRenderingExtension.Render(IReport report, StreamProvider streamProvider, NameValueCollection settings)
		{
			if (report == null)
				throw new ArgumentNullException("report");
			if (streamProvider == null)
				throw new ArgumentNullException("streamProvider");

			settings = settings ?? new NameValueCollection();
			var pdfSettings = new PdfSharpSettings(settings);

			try
			{
				var primaryStream = streamProvider.GetPrimaryStream() ?? streamProvider.CreatePrimaryStream("application/pdf", ".pdf");
				using (var outputStream = primaryStream.OpenStream())
				using (var generator = new PdfGenerator(outputStream, pdfSettings.EmbedFonts))
				{
					var renderingCore = new DocumentRenderer(report, generator);
					int startPage = pdfSettings.StartPage;
					int endPage = Math.Max(startPage, pdfSettings.EndPage);
					if (startPage <= 0)
						startPage = 1;
					if (endPage <= 0)
						endPage = int.MaxValue;
					renderingCore.Render(pdfSettings.Target, false, new ImageRenderers.PageControl.PageControllerSettings() { EndPage = endPage, StartPage = startPage });
				}
			}
			catch
			{
				streamProvider.CleanUpOnError();
				throw;
			}
		}

		#endregion

		#region IConfigurable implementation

		/// <summary>
		/// Gets settings.
		/// </summary>
		ISettings IConfigurable.GetSupportedSettings()
		{
			return new PdfSharpSettings();
		}

		/// <summary>
		/// Gets settings.
		/// </summary>
		ISettings IConfigurable.GetSupportedSettings(bool fpl)
		{
			return new PdfSharpSettings();
		}

		#endregion
	}
}

File: Advanced/PageAndRDL/CustomPdfExport/VB.NET/PdfRendering/Pdf/PdfSharpRenderingExtension.vb
Imports GrapeCity.ActiveReports.Rendering.Export
Imports GrapeCity.ActiveReports.Extensibility.Rendering.IO
Imports GrapeCity.ActiveReports.Extensibility.Rendering.Components
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports System.Collections.Specialized
Imports System

''' <summary>
''' <see cref="IRenderingExtension" /> implementation used to create a PDF.
''' </summary>
Public NotInheritable Class PdfSharpRenderingExtension
	Implements IRenderingExtension
	Implements IConfigurable
#Region "IRenderingExtension implementation"

	''' <summary>
	''' Renders report.
	''' </summary>
	Sub Render(report As IReport, streamProvider As StreamProvider) Implements IRenderingExtension.Render
		CType(Me, IRenderingExtension).Render(report, streamProvider, Nothing)
	End Sub

	''' <summary>
	''' Renders report.
	''' </summary>
	Sub Render(report As IReport, streamProvider As StreamProvider, settings As NameValueCollection) Implements IRenderingExtension.Render
		If report Is Nothing Then
			Throw New ArgumentNullException("report")
		End If
		If streamProvider Is Nothing Then
			Throw New ArgumentNullException("streamProvider")
		End If

		settings = If(settings, New NameValueCollection())
		Dim pdfSettings = New PdfSharpSettings(settings)

		Try
			Dim primaryStream = If(streamProvider.GetPrimaryStream(), streamProvider.CreatePrimaryStream("application/pdf", ".pdf"))
			Using outputStream = primaryStream.OpenStream()
				Using generator As New PdfGenerator(outputStream, pdfSettings.EmbedFonts)
					Dim renderingCore As New DocumentRenderer(report, generator)
					Dim startPage As Integer = pdfSettings.StartPage
					Dim endPage As Integer = Math.Max(startPage, pdfSettings.EndPage)
					If startPage <= 0 Then
						startPage = 1
					End If
					If endPage <= 0 Then
						endPage = Integer.MaxValue
					End If
					renderingCore.Render(pdfSettings.Target, False, New ImageRenderers.PageControl.PageControllerSettings() With {
						.EndPage = endPage,
						.StartPage = startPage
					})
				End Using
			End Using
		Catch
			streamProvider.CleanUpOnError()
			Throw
		End Try
	End Sub

#End Region

#Region "IConfigurable implementation"

	''' <summary>
	''' Gets settings.
	''' </summary>
	Function GetSupportedSettings() As ISettings Implements IConfigurable.GetSupportedSettings
		Return New PdfSharpSettings()
	End Function

	''' <summary>
	''' Gets settings.
	''' </summary>
	Function GetSupportedSettings(fpl As Boolean) As ISettings Implements IConfigurable.GetSupportedSettings
		Return New PdfSharpSettings()
	End Function

#End Region
End Class
